The ``lazy_matrix_mul`` func is found in the ``101-lazy_matrix_mul`` module
======================================================================

The examples below demonstrates the use of the ``lazy_matrix_mul`` function
-----------------------------------------------------------------------

The ``lazy_matrix_mul`` multiplies two matrices
A matrix is a list of lists

Args:

Receives a matrix as first arg, and another matrix as second arg

Return: It returns a result of multiplication

First import the function from the module
``lazy_matrix_mul`` from the ``101-lazy_matrix_mul`` module:

::

    >>> lazy_matrix_mul = __import__("101-lazy_matrix_mul").lazy_matrix_mul

    
    ---- A sample Usage ----
    >>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]))
    [[7, 10], [15, 22]]
    >>> print(lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]]))
    [[13, 16]]

---------------------------------------
Check for presence of module's docstring
----------------------------------------

::

    >>> module_doc = __import__("101-lazy_matrix_mul").__doc__
    >>> len(module_doc) > 1
    True

---------------------------------------
Check for presence of function's docstring
----------------------------------------

::

    >>> func_doc = __import__("101-lazy_matrix_mul").lazy_matrix_mul.__doc__
    >>> len(func_doc) > 1
    True

=============------------------------============
EXCEPTION TEST ON MATRIX
=============------------------------============
CASE 1: <ARG-1> is not a matrix
-------------------------------
::

    >>> lazy_matrix_mul([[1, 2], [3, 'a']], [[1], [3]])
    Traceback (most recent call last):
       ...
    TypeError: m_a should contain only integers or floats

    >>> lazy_matrix_mul([[1, 2], [3]], [[1], [3]])
    Traceback (most recent call last):
       ...
    TypeError: each row of m_a must be of the same size

    >>> lazy_matrix_mul([[1, 2], [3, 2]], [[1, 2], [3]])
    Traceback (most recent call last):
       ...
    TypeError: each row of m_b must be of the same size

    >>> lazy_matrix_mul([], [[1], [3]])
    Traceback (most recent call last):
       ...
    ValueError: m_a can't be empty

-------------------------------------
CORRECT OUPUT TESTS:
-------------------------------------

    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[5, 6], [7, 8]])
    [[19, 22], [43, 50]]
    
    >>> lazy_matrix_mul([[1.5, 2.5], [3.5, 4.5]], [[5.5, 6.5], [7.5, 8.5]])
    [[27.0, 31.0], [53.0, 61.0]]

    >>> lazy_matrix_mul([[1, 2], [3, 'a']], [[1], [3]])
    Traceback (most recent call last):
       ...
    TypeError: m_a should contain only integers or floats

    >>> lazy_matrix_mul([[1, 2], [3, 2]], [[1, 3], [3, 's']])
    Traceback (most recent call last):
       ...
    TypeError: m_b should contain only integers or floats

    >>> lazy_matrix_mul("m_a string", [[1], [3]])
    Traceback (most recent call last):
       ...
    TypeError: m_a must be a list

    >>> lazy_matrix_mul([[1, 2], [3, 2]], "m_b string")
    Traceback (most recent call last):
       ...
    TypeError: m_b must be a list

    >>> lazy_matrix_mul([1, 23, 38], [[1, 3], [3, 2], [22, 3]])
    Traceback (most recent call last):
       ...
    TypeError: m_a must be a list of lists
    
    >>> lazy_matrix_mul([[1, 3], [3, 2], [22, 3]], [2, 4, 6])
    Traceback (most recent call last):
       ...
    TypeError: m_b must be a list of lists

    >>> lazy_matrix_mul([[1, 2], [3]], [[1], [3]])
    Traceback (most recent call last):
       ...
    TypeError: each row of m_a must be of the same size

    >>> lazy_matrix_mul([[1, 3], [3, 2], [22, 3]], [])
    Traceback (most recent call last):
       ...
    ValueError: m_b can't be empty

    >>> lazy_matrix_mul([], [[1], [3]])
    Traceback (most recent call last):
       ...
    ValueError: m_a can't be empty

    >>> lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]])
    [[13, 16]]

    >>> lazy_matrix_mul([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10], [11, 12]])
    [[58, 64], [139, 154]]

    >>> lazy_matrix_mul([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10]])
    Traceback (most recent call last):
       ...
    ValueError: m_a and m_b can't be multiplied

CASE MISSING ARGUMENTS:

    >> print(lazy_matrix_mul([[1, 2]])) # doctest: +NORMALIZE_WHITESPACE
    Traceback (most recent call last):
       ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

    >> print(lazy_matrix_mul()) # doctest: +NORMALIZE_WHITESPACE
    Traceback (most recent call last):
       ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
